File: remote/src/main/java/com/kelsos/mbrc/repository/library/CoverRepositoryImpl.java
package com.kelsos.mbrc.repository.library;

import com.kelsos.mbrc.RemoteDatabase;
import com.kelsos.mbrc.dao.CoverDao;
import com.kelsos.mbrc.dao.TrackDao_Table;
import com.raizlabs.android.dbflow.runtime.TransactionManager;
import com.raizlabs.android.dbflow.sql.language.SQLite;
import com.raizlabs.android.dbflow.structure.BaseModel;
import java.util.List;
import rx.Observable;

public class CoverRepositoryImpl implements CoverRepository {
  @Override public Observable<List<CoverDao>> getPageObservable(int offset, int limit) {
    return Observable.defer(() -> Observable.just(getPage(offset, limit)));
  }

  @Override public Observable<List<CoverDao>> getAllObservable() {
    return Observable.defer(() -> Observable.just(getAll()));
  }

  @Override public List<CoverDao> getPage(int offset, int limit) {
    return SQLite.select()
        .from(CoverDao.class)
        .limit(limit)
        .offset(offset)
        .queryList();
  }

  @Override public List<CoverDao> getAll() {
    return SQLite.select()
        .from(CoverDao.class)
        .queryList();
  }

  @Override public CoverDao getById(long id) {
    return SQLite.select().from(CoverDao.class).where(TrackDao_Table.id.eq(id)).querySingle();
  }

  @Override public void save(List<CoverDao> items) {
    TransactionManager.transact(RemoteDatabase.NAME, () -> {
      Observable.from(items).forEach(BaseModel::save);
    });
  }

  @Override public void save(CoverDao item) {
    item.save();
  }

  @Override public long count() {
    return SQLite.selectCountOf().from(CoverDao.class).count();
  }
}

File: mbrc-lib/src/main/kotlin/com/kelsos/mbrc/repository/library/CoverRepositoryImpl.kt
package com.kelsos.mbrc.repository.library

import com.kelsos.mbrc.RemoteDatabase
import com.kelsos.mbrc.dao.CoverDao
import com.kelsos.mbrc.dao.TrackDao_Table
import com.raizlabs.android.dbflow.runtime.TransactionManager
import com.raizlabs.android.dbflow.sql.language.SQLite
import com.raizlabs.android.dbflow.structure.BaseModel
import rx.Observable
import rx.functions.Action1

class CoverRepositoryImpl : CoverRepository {
    override fun getPageObservable(offset: Int, limit: Int): Observable<List<CoverDao>> {
        return Observable.defer { Observable.just(getPage(offset, limit)) }
    }

    override fun getAllObservable(): Observable<List<CoverDao>> = Observable.defer {
        Observable.just(getAll())
    }

    override fun getPage(offset: Int, limit: Int): List<CoverDao> {
        return SQLite.select()
                .from(CoverDao::class.java)
                .limit(limit)
                .offset(offset)
                .queryList()
    }

    override fun getAll(): List<CoverDao> = SQLite.select()
            .from(CoverDao::class.java)
            .queryList()

    override fun getById(id: Long): CoverDao {
        return SQLite.select()
                .from(CoverDao::class.java)
                .where(TrackDao_Table.id.eq(id))
                .querySingle()
    }


    override fun save(item: CoverDao) {
        item.save()
    }

    override fun count(): Long {
        return SQLite.selectCountOf()
                .from(CoverDao::class.java)
                .count()
    }

    override fun save(items: List<CoverDao>) {
        TransactionManager.transact(RemoteDatabase.NAME) {
            Observable.from(items).forEach({ it.save() })
        }
    }
}
